### DISCUSSION TITLE
[Give your opinion] The way I fixed "Expected server HTML to contain a matching <div> in <div>."

### DISCUSSION THREAD

**Discussion by chiefGui** [vercel/next.js#35560] [Help]
_A TL;DR can be read at the bottom of the post._

I mistakenly and unintentionally fixed the **"Expected server HTML to contain a matching \<div> in \<div>"** in a peculiar way. And by "fixed", I think I mean I bypassed it? Let me show you.

At `_app.tsx`:

```tsx
const Await = ({ children }) => {
  const [isMounted, setIsMounted] = useState(false)
  
  useEffect(() => {
    setIsMounted(true)
  }, [])
  
  if (!isMounted) {
    return null
  }
  
  return children
}

const MyApp = ({ Component }) => {
  return (
    <Await>
      <Component />
    </Await>
  )
}
```

Pretty much, this is a common way to render stuff client-side only ([i.e.](https://gist.github.com/meotimdihia/9faec94a4b223932143cd81a19058f05)) that I bet you've already seen before. But let me give you some extra context on the code above.

### Backstory

Back in 2021, we started writing an app. And [to avoid FOUT](https://dev.to/fyfirman/how-to-fix-fout-flash-of-unstyled-text-in-react-1dl1), I wrote a comparable script (instead of `Await`, it was called `AwaitFonts`) that did the same thing, except that instead of `setIsMounted(true)` at the initial render, I was `setIsMounted(true)` when the font was properly loaded. Again: the _only_ difference between the above demo (`Await`) and our real world situation (`AwaitFonts`) was the time `setIsMounted(true)` was called. Anyways, writing and implementing this `AwaitFonts` was the very first thing I did in this app I mentioned.

So I moved on. I spent months working on this app. Our company released it; it received several patches; customers using and all. We had a blast. **Until today.**

Time for a refactor! We were no longer in need of `AwaitFonts` which led me to remove that piece of code and unwrap the `<Component />` at `_app.tsx` from the `<AwaitFonts />` component. **_Boom! Riot!_** I cannot even describe. Everything seemed broken. At first, I didn't relate the issues I was facing (mostly broken page UI due to `Expected server HTML to contain a matching \<div> in \<div>`) to `<AwaitFonts />` until I had a vision, reminding me about it.

Then I wrapped `<Component />` back, but this time with the `<Await />` you see above instead of the `<AwaitFonts />`.

_Alright, phew._ Regardless of `Await` or `AwaitFonts`, what actually did the trick was a mere re-render when the component was first mounted.

Now...

### TL;DR

```tsx
const [mounted, setMounted] = useState(false)

useEffect(() => {
  setMounted(true)
, [])

return <div>{mounted ? 'client' : 'both'}</div>
```

I usually read solutions like the above when we want to render something client-side only. My question is: **what are we losing by wrapping the entire application when `mounted === true`?**

**Comment by Dragate** [comment]
How does your application perform SEO-wise? Since, on first render, the DOM is empty that must mean the crawler does not "see" anything (initially). Google crawler waits for JS to execute, but not for long. Depending on the complexity of your application it might be enough. From my point of view it looks like a gamble, unless this application does not care about SEO.

**Reply by icyJoseph** [reply]
This.

I'd say that, while this is a valid approach to prevent divergent HTML, it is worth investigating the source of the divergence. Is it a Math.random call, a device width check on Server context? 

If it's really not possible to remove the source of divergence, then some fallback elements in combination with CSS, possibly with visibility hidden (not display none) to reduce layout shift, or plain skipping rendering of those sections and pay the layout shift price, ought to be considered.

At this point you might as well use `dynamic` with `ssr` set to false, and skip creating wrapper components, https://nextjs.org/docs/advanced-features/dynamic-import#with-no-ssr.

**Reply by chiefGui** [reply]
Thank you @Dragate and @icyJoseph. You two guys are right&mdash;we never felt SEO problems because this application in particular doesn't care much about it. If we would spend time with SEO, then I'd have had realised.

Anyways, yes, I'm marching more towards the dynamic import approach&mdash;thank you for emphasizing it!
